File: solution/083RemoveDuplicatesFromSortedList/083RemoveDuplicatesFromSortedList.go
package main

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
	dummy := ListNode{0, head}
	p := dummy.Next
	var q *ListNode
	if p != nil {
		q = p.Next
	} else {
		return nil
	}
	for q != nil {
		if p.Val == q.Val {
			q = q.Next
			p.Next = q
		} else {
			q = q.Next
			p = p.Next
		}
	}
	p.Next = q

	return dummy.Next
}

func main() {

}

File: solution/083RemoveDuplicatesFromSortedList/083RemoveDuplicatesFromSortedList.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = ListNode{val:0, next:head};
        let mut p = dummy.next.as_mut();
        while let Some(q) = p {
            while let Some(r) = q.next.as_mut() {
                if r.val != q.val {
                    break;
                }
                q.next = r.next.take();
            }
            p = q.next.as_mut();
        }
        
        dummy.next
    }
}
